File: backend/app/api/grid.py
import math

from fastapi import APIRouter

from app.services.geocode_core import current_grid_version
from app.services.geometry import lonlat_to_xy, xy_to_lonlat

router = APIRouter(prefix="/v1", tags=["grid"])

MAX_GRID_LINES = 1200
MIN_GRID_ZOOM = 18


@router.get("/grid-version/current")
def current() -> dict:
    return current_grid_version()


@router.get("/grid/viewport")
def viewport_grid(west: float, south: float, east: float, north: float, zoom: float) -> dict:
    if zoom < MIN_GRID_ZOOM:
        return {"visible": False, "reason": "zoom_too_low", "grid": None}

    grid = current_grid_version()
    cell_size_m = grid["cell_size_m"]
    origin_x = grid["origin_x"]
    origin_y = grid["origin_y"]

    corners = [
        lonlat_to_xy(west, south),
        lonlat_to_xy(west, north),
        lonlat_to_xy(east, south),
        lonlat_to_xy(east, north),
    ]
    xs = [point[0] for point in corners]
    ys = [point[1] for point in corners]
    min_x = min(xs)
    max_x = max(xs)
    min_y = min(ys)
    max_y = max(ys)

    x_min_index = math.floor((min_x - origin_x) / cell_size_m) - 1
    x_max_index = math.ceil((max_x - origin_x) / cell_size_m) + 1
    y_min_index = math.floor((min_y - origin_y) / cell_size_m) - 1
    y_max_index = math.ceil((max_y - origin_y) / cell_size_m) + 1

    vertical_count = x_max_index - x_min_index + 1
    horizontal_count = y_max_index - y_min_index + 1
    line_count = vertical_count + horizontal_count
    if line_count > MAX_GRID_LINES:
        return {"visible": False, "reason": "too_many_lines", "line_count": line_count}

    features: list[dict] = []
    for x_index in range(x_min_index, x_max_index + 1):
        x = origin_x + x_index * cell_size_m
        start_lon, start_lat = xy_to_lonlat(x, min_y)
        end_lon, end_lat = xy_to_lonlat(x, max_y)
        features.append(
            {
                "type": "Feature",
                "geometry": {"type": "LineString", "coordinates": [[start_lon, start_lat], [end_lon, end_lat]]},
                "properties": {"kind": "vertical"},
            }
        )

    for y_index in range(y_min_index, y_max_index + 1):
        y = origin_y + y_index * cell_size_m
        start_lon, start_lat = xy_to_lonlat(min_x, y)
        end_lon, end_lat = xy_to_lonlat(max_x, y)
        features.append(
            {
                "type": "Feature",
                "geometry": {"type": "LineString", "coordinates": [[start_lon, start_lat], [end_lon, end_lat]]},
                "properties": {"kind": "horizontal"},
            }
        )

    return {
        "visible": True,
        "reason": None,
        "grid_version": grid["version"],
        "cell_size_m": cell_size_m,
        "line_count": line_count,
        "grid": {"type": "FeatureCollection", "features": features},
    }

File: web/src/grid.ts
import proj4 from "proj4";

proj4.defs("EPSG:32648", "+proj=utm +zone=48 +datum=WGS84 +units=m +no_defs +type=crs");

export type GridMetadata = {
  version: string;
  crs: "EPSG:32648" | string;
  cell_size_m: number;
  origin_x: number;
  origin_y: number;
};

export type LngLatBoundsLike = {
  getWest(): number;
  getSouth(): number;
  getEast(): number;
  getNorth(): number;
};

export type ViewportGridResult = {
  data: GeoJSON.FeatureCollection<GeoJSON.LineString>;
  hiddenReason: string | null;
  hiddenCode: "zoom_too_low" | "too_many_lines" | "projection_error" | null;
  verticalLineCount: number;
  horizontalLineCount: number;
  lineCount: number;
};

export const GRID_MIN_ZOOM = 18;
export const MAX_GRID_LINES = 500;

type Point = [number, number];

export function gridIndex(value: number, origin: number, cellSize: number): number {
  return Math.floor((value - origin) / cellSize);
}

export function viewportGridRange(bounds: LngLatBoundsLike, grid: GridMetadata) {
  const corners = boundsToUtmCorners(bounds);
  const xs = corners.map(([x]) => x);
  const ys = corners.map(([, y]) => y);
  return {
    minXIndex: gridIndex(Math.min(...xs), grid.origin_x, grid.cell_size_m),
    maxXIndex: Math.ceil((Math.max(...xs) - grid.origin_x) / grid.cell_size_m),
    minYIndex: gridIndex(Math.min(...ys), grid.origin_y, grid.cell_size_m),
    maxYIndex: Math.ceil((Math.max(...ys) - grid.origin_y) / grid.cell_size_m),
    minX: Math.min(...xs),
    maxX: Math.max(...xs),
    minY: Math.min(...ys),
    maxY: Math.max(...ys),
  };
}

export function viewportGridLines(
  bounds: LngLatBoundsLike,
  zoom: number,
  grid: GridMetadata,
  maxLines = MAX_GRID_LINES,
): ViewportGridResult {
  if (zoom < GRID_MIN_ZOOM) return hiddenGrid("Zoom in to see the 3m grid", "zoom_too_low");

  let range: ReturnType<typeof viewportGridRange>;
  try {
    range = viewportGridRange(bounds, grid);
  } catch {
    return hiddenGrid("Zoom in to see the 3m grid", "projection_error");
  }
  const verticalLineCount = range.maxXIndex - range.minXIndex + 1;
  const horizontalLineCount = range.maxYIndex - range.minYIndex + 1;
  const lineCount = verticalLineCount + horizontalLineCount;
  if (lineCount > maxLines) {
    return {
      ...hiddenGrid("Zoom in to see the 3m grid", "too_many_lines"),
      verticalLineCount,
      horizontalLineCount,
      lineCount,
    };
  }

  const features: GeoJSON.Feature<GeoJSON.LineString>[] = [];
  for (let xIndex = range.minXIndex; xIndex <= range.maxXIndex; xIndex += 1) {
    const x = grid.origin_x + xIndex * grid.cell_size_m;
    features.push(lineFeature(utmToLngLat([x, range.minY]), utmToLngLat([x, range.maxY])));
  }
  for (let yIndex = range.minYIndex; yIndex <= range.maxYIndex; yIndex += 1) {
    const y = grid.origin_y + yIndex * grid.cell_size_m;
    features.push(lineFeature(utmToLngLat([range.minX, y]), utmToLngLat([range.maxX, y])));
  }

  return { data: { type: "FeatureCollection", features }, hiddenReason: null, hiddenCode: null, verticalLineCount, horizontalLineCount, lineCount };
}

function boundsToUtmCorners(bounds: LngLatBoundsLike): Point[] {
  return [
    lngLatToUtm([bounds.getWest(), bounds.getSouth()]),
    lngLatToUtm([bounds.getWest(), bounds.getNorth()]),
    lngLatToUtm([bounds.getEast(), bounds.getSouth()]),
    lngLatToUtm([bounds.getEast(), bounds.getNorth()]),
  ];
}

function lngLatToUtm(point: Point): Point {
  return proj4("EPSG:4326", "EPSG:32648", point) as Point;
}

function utmToLngLat(point: Point): Point {
  return proj4("EPSG:32648", "EPSG:4326", point) as Point;
}

function lineFeature(start: Point, end: Point): GeoJSON.Feature<GeoJSON.LineString> {
  return { type: "Feature", properties: {}, geometry: { type: "LineString", coordinates: [start, end] } };
}

function hiddenGrid(hiddenReason: string, hiddenCode: ViewportGridResult["hiddenCode"]): ViewportGridResult {
  return { data: { type: "FeatureCollection", features: [] }, hiddenReason, hiddenCode, verticalLineCount: 0, horizontalLineCount: 0, lineCount: 0 };
}
